File: services/order_status.py
"""
Single source of truth for order status strings on the backend.
Frontend mirror lives in `Barista Front End/src/constants/orderStatus.js`
— keep in sync.

THE BUG WE'RE PREVENTING
------------------------
Status used to be a literal string at 100+ call sites. Some spelled
'in-progress' (hyphen), some 'in_progress' (underscore). Filters that
compared with the wrong spelling silently dropped orders.

USAGE
-----
    from services.order_status import OrderStatus, is_in_progress

    if order_row.status == OrderStatus.PENDING:
        ...

    if is_in_progress(order_row):
        ...
"""
from __future__ import annotations


class OrderStatus:
    """Canonical status strings used in the orders.status DB column.

    IMPORTANT:
    - IN_PROGRESS uses a HYPHEN ('in-progress'). The underscore form
      ('in_progress') is legacy drift; tolerate it on READ via
      is_in_progress(), but always WRITE the canonical form.
    - PICKED_UP uses an UNDERSCORE. Don't 'fix' it without a data
      migration — DB rows have this exact value.
    """

    PENDING     = 'pending'
    IN_PROGRESS = 'in-progress'
    COMPLETED   = 'completed'
    PICKED_UP   = 'picked_up'
    CANCELLED   = 'cancelled'


# Set form for `in` checks.
KNOWN_STATUSES = {
    OrderStatus.PENDING, OrderStatus.IN_PROGRESS,
    OrderStatus.COMPLETED, OrderStatus.PICKED_UP, OrderStatus.CANCELLED,
}


def _norm(s):
    return (s or '').strip().lower() if isinstance(s, str) else ''


def _status_of(order_or_status):
    """Accept either a row/dict with .status / ['status'] or a raw string."""
    if isinstance(order_or_status, str):
        return _norm(order_or_status)
    if isinstance(order_or_status, dict):
        return _norm(order_or_status.get('status'))
    return _norm(getattr(order_or_status, 'status', None))


def is_pending(order_or_status) -> bool:
    return _status_of(order_or_status) == 'pending'


def is_in_progress(order_or_status) -> bool:
    """Tolerates the underscore drift for legacy callers/rows."""
    s = _status_of(order_or_status)
    return s in ('in-progress', 'in_progress', 'inprogress')


def is_completed(order_or_status) -> bool:
    s = _status_of(order_or_status)
    return s in ('completed', 'complete', 'done')


def is_picked_up(order_or_status) -> bool:
    s = _status_of(order_or_status)
    return s in ('picked_up', 'picked-up', 'pickedup', 'collected')


def is_cancelled(order_or_status) -> bool:
    s = _status_of(order_or_status)
    return s in ('cancelled', 'canceled', 'cancel')


def is_active(order_or_status) -> bool:
    """Not yet collected or cancelled — useful for queue counts."""
    return is_pending(order_or_status) or is_in_progress(order_or_status)


def canonical_status(order_or_status):
    """Map a status to its canonical OrderStatus value, normalising
    legacy spellings. Returns None if unrecognised."""
    if is_pending(order_or_status):     return OrderStatus.PENDING
    if is_in_progress(order_or_status): return OrderStatus.IN_PROGRESS
    if is_completed(order_or_status):   return OrderStatus.COMPLETED
    if is_picked_up(order_or_status):   return OrderStatus.PICKED_UP
    if is_cancelled(order_or_status):   return OrderStatus.CANCELLED
    return None


# For SQL `IN` clauses where you want the underscore drift accepted.
IN_PROGRESS_SQL_VARIANTS = ('in-progress', 'in_progress')
PICKED_UP_SQL_VARIANTS   = ('picked_up', 'picked-up')

File: Barista Front End/src/constants/orderStatus.js
// constants/orderStatus.js
//
// Single source of truth for order status strings on the frontend.
// Backend mirror lives in services/order_status.py — keep in sync.
//
// THE BUG WE'RE PREVENTING
// ------------------------
// Order status used to be a literal string scattered through 100+
// call sites. Some places spelled 'in-progress' (hyphen), some
// spelled 'in_progress' (underscore). Filters that compared with the
// wrong spelling silently dropped orders.
//
// USAGE
// -----
//   import { ORDER_STATUS, isInProgress } from './orderStatus';
//
//   if (order.status === ORDER_STATUS.PENDING) { ... }
//   if (isInProgress(order)) { ... }
//
//   // For new code, prefer the helpers over raw === comparisons.

/**
 * Canonical status strings. These are the values stored in the DB
 * `orders.status` column and sent over the wire by the backend.
 *
 * IMPORTANT:
 * - IN_PROGRESS uses a HYPHEN ('in-progress'). The underscore form
 *   ('in_progress') is wrong but appears in some legacy spots and
 *   must be tolerated when READING (use `isInProgress` not `===`).
 * - PICKED_UP uses an UNDERSCORE. Yes, that's inconsistent. Don't
 *   "fix" it without a migration — DB rows have this value.
 */
export const ORDER_STATUS = Object.freeze({
  PENDING:     'pending',
  IN_PROGRESS: 'in-progress',
  COMPLETED:   'completed',
  PICKED_UP:   'picked_up',
  CANCELLED:   'cancelled',
});

// Set form for quick `in` checks. Object.values() works too but a Set
// is constant-time and more readable as `KNOWN_STATUSES.has(...)`.
export const KNOWN_STATUSES = new Set(Object.values(ORDER_STATUS));

/**
 * Tolerant matchers — accept the canonical form AND the known legacy
 * spellings. Use these for ANY filter/branch that reads order.status
 * coming from older clients, mock data, or test fixtures.
 *
 * If you're writing a NEW status, write the canonical form
 * (ORDER_STATUS.*) — these helpers exist for reads, not writes.
 */
const _norm = (s) => (s || '').toString().toLowerCase().trim();

export const isPending = (orderOrStatus) => {
  const s = _norm(typeof orderOrStatus === 'string' ? orderOrStatus : orderOrStatus?.status);
  return s === 'pending';
};

export const isInProgress = (orderOrStatus) => {
  const s = _norm(typeof orderOrStatus === 'string' ? orderOrStatus : orderOrStatus?.status);
  // Tolerate the underscore drift.
  return s === 'in-progress' || s === 'in_progress' || s === 'inprogress';
};

export const isCompleted = (orderOrStatus) => {
  const s = _norm(typeof orderOrStatus === 'string' ? orderOrStatus : orderOrStatus?.status);
  return s === 'completed' || s === 'complete' || s === 'done';
};

export const isPickedUp = (orderOrStatus) => {
  const s = _norm(typeof orderOrStatus === 'string' ? orderOrStatus : orderOrStatus?.status);
  return s === 'picked_up' || s === 'picked-up' || s === 'pickedup' || s === 'collected';
};

export const isCancelled = (orderOrStatus) => {
  const s = _norm(typeof orderOrStatus === 'string' ? orderOrStatus : orderOrStatus?.status);
  return s === 'cancelled' || s === 'canceled' || s === 'cancel';
};

/**
 * 'Active' = not yet collected/cancelled. Useful for queue counts.
 */
export const isActive = (orderOrStatus) =>
  isPending(orderOrStatus) || isInProgress(orderOrStatus);

/**
 * Map an order to its canonical status string, normalising legacy
 * spellings to the ORDER_STATUS values. Returns undefined if the
 * status doesn't match any known form.
 */
export const canonicalStatus = (orderOrStatus) => {
  if (isPending(orderOrStatus))     return ORDER_STATUS.PENDING;
  if (isInProgress(orderOrStatus))  return ORDER_STATUS.IN_PROGRESS;
  if (isCompleted(orderOrStatus))   return ORDER_STATUS.COMPLETED;
  if (isPickedUp(orderOrStatus))    return ORDER_STATUS.PICKED_UP;
  if (isCancelled(orderOrStatus))   return ORDER_STATUS.CANCELLED;
  return undefined;
};

/**
 * Human-friendly label for the status. Use in UI text.
 */
export const STATUS_LABELS = Object.freeze({
  [ORDER_STATUS.PENDING]:     'Pending',
  [ORDER_STATUS.IN_PROGRESS]: 'In progress',
  [ORDER_STATUS.COMPLETED]:   'Ready',
  [ORDER_STATUS.PICKED_UP]:   'Collected',
  [ORDER_STATUS.CANCELLED]:   'Cancelled',
});

export const labelFor = (orderOrStatus) => {
  const c = canonicalStatus(orderOrStatus);
  return c ? STATUS_LABELS[c] : 'Unknown';
};

export default ORDER_STATUS;
